Add Redux integration recipe (#26)
moxystudio/react-native-with-moxy: docusaurus/docs/recipes/redux.mdx

---
id: redux
title: Setting up Redux
sidebar_label: Redux
---

import useBaseUrl from '@docusaurus/useBaseUrl';

<img src={ useBaseUrl('recipes-assets/setting-up-redux/redux-logo.svg') } alt="Redux" width="300" />

In this recipe, you will be guided through the process of setting up a disk persistance-enabled Redux store in your React Native app.

> > ℹ️ It is assumed you have understanding of Redux concepts, how it works and knowledge of the most common tools revolving Redux's ecossystem. In case you don't, first please go through the [official introductory document](https://redux.js.org/basics/basic-tutorial) before proceeding. We also recommend you to also read the Advanced Tutorial material, otherwise it might be difficult for you to follow this recipe.

Reviewer: ```suggestion
> ℹ️ It is assumed you have an understanding of Redux concepts, how it works, and knowledge of the most common tools revolving Redux's ecosystem. In case you don't, first please go through the [official introductory document](https://redux.js.org/basics/basic-tutorial) before proceeding. We also recommend you read the Advanced Tutorial material, otherwise, it might be difficult for you to follow this recipe.
```

Author: Done ✅

- Maintain different versions of the state and move from an older version to a newer one with migrations.

By default, the whole state tree is persisted. Head to [`redux-persist`](https://github.com/rt2zz/redux-persist) repository to learn more about how Redux store persistance works and how it can be configured and customized.

## 3. Create `StoreProvider`

The next step is to put together Redux's `Provider` and `redux-persist`'s `PersistGate`.

Create the files as shown in the sections below.

#### `src/shared/modules/react-native-store/StoreProvider.js`

```js
import React, { useEffect, useState, useRef } from 'react';
import PropTypes from 'prop-types';
import { Provider } from 'react-redux';
import { PersistGate } from 'redux-persist/integration/react';

const StoreProvider = ({ children, ...props }) => {
    const { current: { store, purge, persist } } = useRef(props);
    const [isPurging, setPurging] = useState(purge && !!store.persistor);
    const childrenFn = (storeReady) => children({ storeReady });

    useEffect(() => {
        if (!persist) {

>             store?.persistor.pause();

Reviewer: If you're setting the store prop as required in the PropTypes, why null check it here?

Author: The `store` is being required but `persistor` is not. However, we're not really expecting to receive a store without persistence capabilities. This is just a safe guard to ensure an error is thrown if persistence is disabled by the developer and the store cannot persist state. This is effectively a no-op.

Author: I think we're missing a `?.` though between `persistor` and `pause()`.

Reviewer: Yes, the `?.` should be between the `persistor` and `pause` 😄

                { childrenFn }
            </PersistGate>
        </Provider>
    );
};

StoreProvider.propTypes = {
    children: PropTypes.func.isRequired,
    persist: PropTypes.bool,
    purge: PropTypes.bool,
    store: PropTypes.shape({
        dispatch: PropTypes.func.isRequired,
        getState: PropTypes.func.isRequired,
        persistor: PropTypes.shape({
            purge: PropTypes.func.isRequired,
            pause: PropTypes.func.isRequired,
        }),
    }).isRequired,
};

StoreProvider.defaultProps = {
    persist: true,
    purge: false,
};

export default StoreProvider;
```

How `StoreProvider` works:
- Persistance is enabled by default. You can choose to disable persistance by setting `persist` prop to `false`. `StoreProvider` does not listen to changes to `persist` after mounting.
- You can choose to purge the persisted state by setting `purge` prop to `true`. `StoreProvider` does not listen to changes to `purge` after mounting.
- The `children` render prop is called with `storeReady` set to `false` while the store is rehydrating. Once rehydrated, `storeReady` changes to `true`.

## 3. Wrap navigation with `StoreProvider`

> Finally, all is left to do is render `StoreProvider` on top of your app's navigation.

Reviewer: ```suggestion
Finally, all that is left to do is render `StoreProvider` on top of your app's navigation.
```

Author: Done ✅

#### `src/shared/state/buildStore.js`

```js
import { createStore, applyMiddleware, combineReducers } from 'redux';
import thunkMiddleware from 'redux-thunk';
import { composeWithDevTools } from 'redux-devtools-extension/developmentOnly';
import { createPersistReducer, createStorePersistor } from './persist';
import { reducer as appReducer, middlewares as appMiddlewares } from './app';

const reducers = {
    app: appReducer,
};

const initialState = {};

const modules = {};

const buildStore = (reducerFn = combineReducers) => {
    const middlewares = [
        thunkMiddleware.withExtraArgument(modules),
        appMiddlewares.logVersionChangeMiddleware,
    ];
    const enhancer = composeWithDevTools(
        applyMiddleware(...middlewares),
    );
    const rootReducer = reducerFn(reducers);
    const store = createStore(rootReducer, initialState, enhancer);

    return store;
};

const buildPersistStore = () => {
    const store = buildStore(createPersistReducer);

    return {
        ...store,
        persistor: createStorePersistor(store),
    };
};

export default buildPersistStore;
```

## 6. Source code

> If'd like, feel free to checkout the [branch](https://github.com/moxystudio/react-native-with-moxy/tree/503a6829b359d8abd018e7ce7f879b4b54a2a0c1) from the boilerplate's repository to see the persistance-enabled Redux store working on both Android and iOS. You can also find tests for [`StoreProvider`](https://github.com/moxystudio/react-native-with-moxy/blob/503a6829b359d8abd018e7ce7f879b4b54a2a0c1/src/shared/modules/react-native-store/StoreProvider.test.js).

Reviewer: ```suggestion
If you would like, feel free to check out the [branch](https://github.com/moxystudio/react-native-with-moxy/tree/503a6829b359d8abd018e7ce7f879b4b54a2a0c1) from the boilerplate's repository to see the persistance-enabled Redux store working on both Android and iOS. You can also find tests for [`StoreProvider`](https://github.com/moxystudio/react-native-with-moxy/blob/503a6829b359d8abd018e7ce7f879b4b54a2a0c1/src/shared/modules/react-native-store/StoreProvider.test.js).
```

Author: Done ✅
